server/src/main/java/org/elasticsearch/script/ScriptedMetricAggContexts.java

```diff
@@ -31,6 +33,25 @@
 import java.util.Map;
 
 public class ScriptedMetricAggContexts {
+    private static final DeprecationLogger DEPRECATION_LOGGER =
+        new DeprecationLogger(Loggers.getLogger(ScriptedMetricAggContexts.class));
+
+    // Public for access from tests
+    public static final String AGG_PARAM_DEPRECATION_WARNING =
+        "params._agg/_aggs for scripted metric aggregations are deprecated, use state/states (not in params) instead. " +
+        "Set system property es.aggregations.enableDeprecatedScriptedMetricAggParam = false to disable this deprecated behavior.";
+
+    public static boolean deprecatedAggParamEnabled() {
+        boolean enabled = Boolean.parseBoolean(
+            System.getProperty("es.aggregations.enableDeprecatedScriptedMetricAggParam", "true"));
+
+        if (enabled) {
+            DEPRECATION_LOGGER.deprecatedAndMaybeLog("enableDeprecatedScriptedMetricAggParams", AGG_PARAM_DEPRECATION_WARNING);
```

Comment:
Unfortunately I've lead you in an incorrect direction. After discussing with others, we want these to only be triggered when the deprecated behavior is used. See #31441 for and example doing this in ScriptDocValues for an idea of what must be done within the aggs code (although I think it will be simpler here, since you can do this check and deprecation outside of script invocation, and thus no need for a doPriv block).